### ISSUE TITLE
Static spec doesn't differentiate between primary and foreign IDs

### ISSUE THREAD

**Issue by botanize** [google/transit#266]
No where in the spec is it clear what the rules are for ID fields. The only "documentation" is in the validators.

For example, it seems reasonable that `trip_id` would only need to be unique within the combination of service and route IDs in the `trips.txt` file (in SQL terms, `PRIMARY KEY (service_id, route_id, trip_id)`). However, validators tell me that `trip_id` is required to be unique within `trips.txt` regardless of the combination of other ID fields (`PRIMARY KEY (trip_id)`), but the other ID fields,  `route_id` and `service_id`, are not expected to be unique. 

Relying on implied rules results in further confusion. For example, the `service_id` ID is the only ID in `calendar_dates.txt`, but is not expected to be unique (`FOREIGN KEY (service_id)`), which you might expect if you assumed that an ID needs to be unique if it's the only ID, or if the ID name is a close match to the file name or file purpose. 

Maybe adapting language from relational databases, e.g. primary ID and foreign ID, would clarify the role of each ID in each file. If there are any examples where there is a multi-column primary key (multiple ID fields used to uniquely identify a row) and that constraint is expected by validators, it should be noted in the file information. 

For example, the spec for stop_times.txt could look like this:

```
### stop_times.txt

File: **Required**

PRIMARY KEY(trip_id, stop_sequence)

|  Field Name | Type | Required | Description |
|  ------ | ------ | ------ | ------ |
|  `trip_id` | foreign ID referencing `trips.trip_id` | **Required** | Identifies a trip.  |

```

**Comment by scmcca** [comment]
This looks good to me! Any thoughts on turning this into a pull request to work through the details?

**Comment by botanize** [comment]
Yep, happy to, but I'm working on getting my employer to sign a contributor agreement first.

**Comment by abyrd** [comment]
Agreed, this is a good idea. I believe GTFS emerged historically as a text dump of a relational database, so relational concepts like primary keys are completely relevant. Especially since GTFS is often used by importing it back into a relational database.

**Comment by IvanVolosyuk** [comment]
Trip id is supposed to be globally unique within a static feed regardless
of the routes they are in. A realtime update can specify only the trip_id
and it should be sufficient to identify the trip uniquely without
additional information.

**Comment by mmorang** [comment]
Actually if you use frequencies.txt, all trips with the same pattern described in frequencies.txt have the same trip_id. So it's unique in that it describes a unique sequence of stops with a defined travel time between stops, but it's not guaranteed to be unique throughout the day. Kind of weird, in my opinion, actually.

**Comment by skinkie** [comment]
> Actually if you use frequencies.txt, all trips with the same pattern described in frequencies.txt have the same trip_id. So it's unique in that it describes a unique sequence of stops with a defined travel time between stops, but it's not guaranteed to be unique throughout the day. Kind of weird, in my opinion, actually.

You can really see where the inspiration for these kind of things came from. This is a very German approach making that requires the departure time and stop to identify a trip. You can not even state it is a _block_ (nothing requires that the frequency is executed consecutively by a single vehicle), it is rather a _pattern_ identification. In my opinion it is a valid question why these things should not be denormalised in unique trips (as the standard loves to do...)

@abyrd what do you think about these things from a travel information perspective. It is fair to have a frequency based approach if the time is by estimation, especially because you can specify the rhythm instead?

**Comment by barbeau** [comment]
I've never been a fan of overloading frequencies.txt with exact_times=1 trips. IMHO it makes the spec much harder to understand and makes discussion of true frequency based trips (i.e., exact_times=0) much harder to discuss because of the ambiguous definition of "frequency". Especially when you add real-time data.

I do understand why it was done, though - to make it easier to hand-edit GTFS in the early days for smaller transit systems (i.e., to increase adoption).

**Comment by IvanVolosyuk** [comment]
What about creating a glossary of terms? One GTFS trip can represent
multiple "trip runs" (or whatever we agree) for the frequency-based trips.
If we have to live with the GTFS definition of the trip for the foreseeable
future. This will make it less ambiguous:
"A realtime update can specify only the trip_id and it should be sufficient
to identify the trip uniquely without additional information."
"A realtime update can specify the trip_id, start_date and start_time, and
it should be sufficient to identify the trip-run uniquely without
additional information."

**Comment by abyrd** [comment]
>> A realtime update can specify only the trip_id and it should be sufficient to identify the trip uniquely without additional information.
>
>Actually if you use frequencies.txt, all trips with the same pattern described in frequencies.txt have the same trip_id.

Also, although a trip_id will uniquely identify a trip within a day, that trip can occur on many different days. If the trip is long enough (especially >24 hours), or if it has been significantly delayed, or if cancellations are known well in advance, it may not be clear which instance of the trip the realtime message refers to. So accounting for the possibility of exact_times frequencies, the unique key for an individual "vehicle journey" is (trip_id, date, first departure time).

**Comment by IvanVolosyuk** [comment]
Agree, and don't see a contradiction for certain definitions of 'trip' and
'vehicle journey'. Thus the need for a glossary to make sure we are talking
about the same concepts. Otherwise, we may spend time arguing about
something because of completely different interpretations.

**Comment by scmcca** [comment]
@abyrd 
>So accounting for the possibility of exact_times frequencies, the unique key for an individual "vehicle journey" is (trip_id, date, first departure time).

Where would "date" get pulled from? Is it meant to be the `service_id`, or an actual date falling within a `calendar.txt` or `calendar_dates.txt` record (associated to the `trip_id` via `service_id`)?

Similarly for first departure time, would these have to be deducted from `frequencies.txt` to generate identifiers for first possible departure times for all vehicle journeys within a (`start_time`, `end_time`) interval?
